File: Linked-Lists/linked_list.cpp
#include <iostream>
using namespace std;

class LinkedList {
    class Node {
       public:
        int value;
        Node *link;

        Node(int value) {
            this->value = value;
        }
    };

   public:
    Node *head;
    int length;

    void traverse() {
        Node *temp = head;
        while (temp != NULL) {
            cout << temp->value << " -> ";
            temp = temp->link;
        }
        cout << "[X]\n";
    }

    void insertNode(int value, int position) {
        // indexing starts at 1
        // if position = 1 : insert at first
        // if position = -1 : insert at last
        // if position = n : insert at index n

        Node *newNode = new Node(value);

        if (position == 1) {
            newNode->link = head;
            head = newNode;
        } else if (position == -1) {
            if (head == NULL) {
                head = newNode;
            } else {
                Node *lastNode = head;
                while (lastNode->link != NULL) {
                    lastNode = lastNode->link;
                }
                lastNode->link = newNode;
            }
        } else {
            if (position < 0 || position > length + 1) {
                cout << "invalid position" << endl;
            } else {
                Node *prevNode = head;
                for (int i = 1; i < position - 1; i++) {
                    prevNode = prevNode->link;
                }
                newNode->link = prevNode->link;
                prevNode->link = newNode;
            }
        }

        length++;
    }

    int deleteNode(int position) {
        // if position = 1 : delete first node
        // if position = n : delete nth node

        if (head == NULL) {
            cout << "empty list" << endl;
            return -1;
        }

        int value;

        if (position == 1) {
            value = head->value;
            head = head->link;
        } else if (position > 1 && position <= length) {
            Node *prevNode = head;
            for (int i = 1; i < position - 1; i++) {
                prevNode = prevNode->link;
            }
            value = prevNode->link->value;
            prevNode->link = prevNode->link->link;
        } else {
            cout << "invalid position" << endl;
            return -1;
        }

        length--;
        return value;
    }

    LinkedList() {
        head = NULL;
        length = 0;
    }
};

int main() {
    LinkedList list;
    list.insertNode(5, 1);
    list.insertNode(10, 1);
    list.insertNode(15, -1);
    list.insertNode(20, 2);
    list.traverse();
    cout << "deleted Node: " << list.deleteNode(2) << endl;
    list.traverse();
    return 0;
}

File: Linked-Lists/linked_list.py
class LinkedList:

    class Node:
        def __init__(self, val=0, next=None):
            self.val = val
            self.next = next
    
    def __init__(self):
        self.length = 0
        self.head = None
        
    def traverse(self):
        temp = self.head
        while temp != None:
            print(temp.val)
            temp = temp.next
        

    def get(self, index: int) -> int:
        if index > self.length - 1 or index < 0:
            return -1
        
        temp = self.head
        while index != 0:
            temp = temp.next
            index -= 1
            
        if temp is not None:
            return temp.val
        else:
            return -1

            
    def reverse(self):
        current = self.head
        prev = None
        while current is not None:
            nextNode = current.next
            current.next = prev
            prev = current
            current = nextNode
        self.head = prev
        


    def addAtHead(self, val: int) -> None:
        node = self.Node(val)
        node.next = self.head
        self.head = node
        
        self.length += 1

    def addAtTail(self, val: int) -> None:
        if self.head is None:
            self.addAtHead(val)
        else:
            node = self.Node(val) 
            temp = self.head
            while temp.next != None:
                temp = temp.next
            temp.next = node
            node.next = None
            
        self.length +=1
            

    def addAtIndex(self, index: int, val: int) -> None:
        if index == 0:
            self.addAtHead(val)
        elif index > self.length or index < 0:
            return
        elif index == self.length:
            self.addAtTail(val)
        else:
            temp = self.head
            while index != 1:
                temp = temp.next
                index -= 1
            node = self.Node(val)
            node.next = temp.next
            temp.next = node
            self.length += 1
            

    def deleteAtIndex(self, index: int) -> None:
        prev = None
        temp = self.head
        
        if index == 0:
            self.head = self.head.next
        elif index > self.length - 1 or index < 0:
            return
        else:
            while index != 0:
                prev = temp
                temp = temp.next
                index -= 1
            prev.next = temp.next
            
        self.length -= 1